File: src/main/java/ch06/observeon/ObserveOnExample1.java
package ch06.observeon;

import static com.google.common.util.concurrent.Uninterruptibles.sleepUninterruptibly;

import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;

import java.util.concurrent.TimeUnit;

public class ObserveOnExample1 {

    public static void main(String[] args) {

        Observable.just("WHISKEY/27653/TANGO", "6555/BRAVO", "232352/5675675/FOXTROT")
                .flatMap(s -> Observable.fromArray(s.split("/")))
                .observeOn(Schedulers.computation())
                .filter(s -> s.matches("[0-9]+"))
                .map(Integer::valueOf)
                .reduce((total, next) -> total + next)
                .subscribe(value ->
                        System.out.printf("Received %d on thread %s", value, Thread.currentThread().getName()));

        sleepUninterruptibly(1, TimeUnit.SECONDS);
    }

}

File: src/main/kotlin/ch06/observeon/observeOn_example1.kt
package ch06.observeon

import ch06.subscribeon.currentThreadName
import com.google.common.util.concurrent.Uninterruptibles.sleepUninterruptibly
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.TimeUnit.SECONDS

/**
 * Shows how to use observeOn in a basic fashion.
 */
fun main(args: Array<String>) {

    Observable.just("WHISKEY/27653/TANGO", "6555/BRAVO", "232352/5675675/FOXTROT")
        .subscribeOn(Schedulers.io())
        .flatMap { Observable.fromIterable(it.split("/")) }
        .observeOn(Schedulers.computation())
        .filter { it -> it.matches(Regex("[0-9]+")) }
        .map { it.toInt() }
        .reduce { total, next -> total + next }
        .subscribe { println("Received $it on thread ${currentThreadName()}") }

    sleepUninterruptibly(1, SECONDS)
}
